exporter/loadbalancingexporter/loadbalancer.go

```diff
@@ -0,0 +1,193 @@
+	oCfg := cfg.(*Config)
+
+	if oCfg.Resolver.DNS != nil && oCfg.Resolver.Static != nil {
+		return nil, errMultipleResolversProvided
+	}
+
+	var res resolver
+	if oCfg.Resolver.Static != nil {
+		var err error
+		res, err = newStaticResolver(oCfg.Resolver.Static.Hostnames)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if oCfg.Resolver.DNS != nil {
+		dnsLogger := params.Logger.With(zap.String("resolver", "dns"))
+
+		var err error
+		res, err = newDNSResolver(dnsLogger, oCfg.Resolver.DNS.Hostname, oCfg.Resolver.DNS.Port)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if res == nil {
+		return nil, errNoResolver
+	}
+
+	return &loadBalancerImp{
+		logger:           params.Logger,
+		res:              res,
+		componentFactory: factory,
+		exporters:        map[string]component.Exporter{},
+	}, nil
+}
+
+func (lb *loadBalancerImp) Start(ctx context.Context, host component.Host) error {
+	lb.res.onChange(lb.onBackendChanges)
+	lb.host = host
+	if err := lb.res.start(ctx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (lb *loadBalancerImp) onBackendChanges(resolved []string) {
+	newRing := newHashRing(resolved)
+
+	if !newRing.equal(lb.ring) {
+		lb.updateLock.Lock()
+		defer lb.updateLock.Unlock()
+
+		lb.ring = newRing
+
+		// TODO: set a timeout?
+		ctx := context.Background()
+
+		// add the missing exporters first
+		lb.addMissingExporters(ctx, resolved)
+		lb.removeExtraExporters(ctx, resolved)
+	}
+}
+
+func (lb *loadBalancerImp) addMissingExporters(ctx context.Context, endpoints []string) {
+	for _, endpoint := range endpoints {
+		endpoint = endpointWithPort(endpoint)
+
+		if _, exists := lb.exporters[endpoint]; !exists {
+			exp, err := lb.componentFactory(ctx, endpoint)
+			if err != nil {
+				lb.logger.Error("failed to create new exporter for endpoint", zap.String("endpoint", endpoint), zap.Error(err))
+				continue
+			}
+
+			if err = exp.Start(ctx, lb.host); err != nil {
+				lb.logger.Error("failed to start new exporter for endpoint", zap.String("endpoint", endpoint), zap.Error(err))
+				continue
+			}
+			lb.exporters[endpoint] = exp
+		}
+	}
+}
+
+func endpointWithPort(endpoint string) string {
+	if !strings.Contains(endpoint, ":") {
+		endpoint = fmt.Sprintf("%s:%s", endpoint, defaultPort)
+	}
+	return endpoint
+}
+
+func (lb *loadBalancerImp) removeExtraExporters(ctx context.Context, endpoints []string) {
+	for existing := range lb.exporters {
+		if !endpointFound(existing, endpoints) {
+			lb.exporters[existing].Shutdown(ctx)
+			delete(lb.exporters, existing)
+		}
+	}
+}
+
+func endpointFound(endpoint string, endpoints []string) bool {
+	for _, candidate := range endpoints {
+		if candidate == endpoint {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (lb *loadBalancerImp) Shutdown(context.Context) error {
+	lb.stopped = true
+	return nil
+}
+
+func (lb *loadBalancerImp) Exporter(traceID pdata.TraceID) (component.Exporter, string, error) {
```

Comment:
This should return only the exporter and error. If you do think that we should have both the exporter and the endpoint to be returned, might be good to wrap them in a custom type.

Reply:
Changed to return only `Exporter`.

[`37c46e6` (#2633)](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2633/commits/37c46e611a2ccc626d1ea74d82e8220b9d85ad35)